File: EmsBMSReports/MainForm.vb
Imports System.Data.Odbc
Imports MactusReportLib.MactusReportLib
Imports MactusReportLib.IndusoftEMSReport
Imports System.ComponentModel

Public Class MainForm

    Private m_nSelectedReportType As ReportType
    Private m_nIntervalMin As Integer = 1
    Private m_nReportID As Integer = 0
    Private m_nReportStatusID As Long = 0

    Private Sub MainForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load

        Try
            'If g_sCurrent_login_UserID.Length > 0 Then
            Dim sError As String = ""

            Dim oTime As DateTime = DateTime.Now
            If ReadDatabaseConnection(sError) = False Then
                MsgBox(sError)
                End
            End If

            bCOnfigureReports.Visible = True
            bConfigureAlarmReports.Visible = True 'g_bEnableConfiguration

            Dim sQuery As String
            sQuery = "UPDATE tbl_reportstatus SET status=4 where status=3"
            ExecuteSQLInDb(sQuery)

            m_oEBOReprots = New MactusReportLib.EBOReport()
            m_oIndusoftReports = New MactusReportLib.NewInduSoftReport()

            bAreaReports.Checked = True
            bAlarmReports.Checked = False
            bEventReport.Checked = False
            If g_bIsBMS = 2 Then
                btnBatteryPercentage.Visible = True
                btnExcursionReport.Visible = True
            End If
            LoadReportsIntoListBox(oReportGrid, ReportType.DataReport)
            cInterval.SelectedIndex = 0
            m_nSelectedReportType = ReportType.DataReport
            oTime = oTime.AddMilliseconds(-oTime.Millisecond)
            oTime = oTime.AddSeconds(-oTime.Second)
            oTime = oTime.AddMinutes(-oTime.Minute)
            oTime = oTime.AddHours(-oTime.Hour)
            oTime = oTime.AddDays(-7)
            tr_FromDate.Value = oTime
            tr_ToDate.Value = oTime.AddDays(1)
            bGenerate.Enabled = False

            ' Dim sQuery As String
            Dim oReader As OdbcDataReader
            sQuery = "SELECT * FROM TBL_DataGroups WHERE grouptype=0 ORDER BY groupid"

            Try
                Using oConnection As New OdbcConnection(g_sConString)
                    oConnection.Open()
                    Dim oCmd As New OdbcCommand(sQuery, oConnection)
                    oReader = oCmd.ExecuteReader()
                    While oReader.Read()
                        cGroup.Items.Add(oReader("groupname"))
                    End While
                    oConnection.Close()
                End Using
            Catch ex As Exception

            End Try

            Try
                cGroup.SelectedIndex = 0
            Catch ex As Exception

            End Try

            UpdateGrid()
            StartThread()


            oTimer.Enabled = True

            '  Else

            '  End If
        Catch ex As Exception

        End Try


    End Sub

    Private Sub bGenerate_Click(sender As Object, e As EventArgs) Handles bGenerate.Click
        If m_nReportID = 0 Then
            MsgBox("No Report Selected")
            Exit Sub
        End If
        oProgress.Value = 0
        Dim nTimeInterval As Integer

        nTimeInterval = Convert.ToInt16(cInterval.Text)

        m_nReportStatusID = InsertNewReportStatusRecord(m_nReportID, tr_FromDate.Value, tr_ToDate.Value, nTimeInterval, "user Name", m_nSelectedReportType)



    End Sub

    Private Sub oReportGrid_CellClick(sender As Object, e As DataGridViewCellEventArgs) Handles oReportGrid.CellClick

        Try
            m_nReportID = oReportGrid.Rows(e.RowIndex).Cells(0).Value
            If m_nSelectedReportType = ReportType.DataReport Then
                cInterval.Text = oReportGrid.Rows(e.RowIndex).Cells(2).Value.ToString
                Label5.Visible = True
                cInterval.Visible = True
                m_nIntervalMin = Convert.ToInt16(oReportGrid.Rows(e.RowIndex).Cells(2).Value)
                AdjustTimeMinuetsToInterval(tr_FromDate.Value)
                AdjustTimeMinuetsToInterval(tr_ToDate.Value)
                bGenerateChart.Enabled = True
            Else
                Label5.Visible = False
                m_nIntervalMin = 1
                bGenerateChart.Enabled = False
            End If
            bGenerate.Enabled = True
        Catch ex As Exception
            m_nReportID = 0
            Label5.Visible = False
            m_nIntervalMin = 1
            bGenerate.Enabled = False
            cInterval.Visible = False
            bGenerateChart.Enabled = False
        End Try

    End Sub

    Private Sub bAreaReports_CheckedChanged(sender As Object, e As EventArgs) Handles bAreaReports.CheckedChanged

        cGroup.Enabled = True
        LoadReportsIntoListBox(oReportGrid, ReportType.DataReport)
        Label5.Visible = False
        m_nSelectedReportType = ReportType.DataReport
    End Sub

    Private Sub bAlarmReports_CheckedChanged(sender As Object, e As EventArgs) Handles bAlarmReports.CheckedChanged
        cGroup.SelectedIndex = -1
        cGroup.Enabled = False
        LoadReportsIntoListBox(oReportGrid, ReportType.AlarmReport)
        Label5.Visible = False
        m_nIntervalMin = 1
        m_nSelectedReportType = ReportType.AlarmReport
    End Sub

    Private Sub bEventReport_CheckedChanged(sender As Object, e As EventArgs) Handles bEventReport.CheckedChanged
        cGroup.SelectedIndex = -1
        cGroup.Enabled = False
        LoadReportsIntoListBox(oReportGrid, ReportType.EventReport)
        Label5.Visible = False
        m_nIntervalMin = 1
        m_nSelectedReportType = ReportType.EventReport
    End Sub

    Private Sub AdjustTimeMinuetsToInterval(ByRef oDateTime As DateTime)
        Try
            Dim nTemp As Integer

            If m_nIntervalMin = 1 Then
                Exit Sub
            End If
            nTemp = oDateTime.Minute / m_nIntervalMin

            nTemp = oDateTime.Minute - nTemp * m_nIntervalMin

            oDateTime = oDateTime.AddMinutes(-nTemp)
        Catch ex As Exception

        End Try

    End Sub

    Private Sub tr_FromDate_LostFocus(sender As Object, e As EventArgs) Handles tr_FromDate.LostFocus
        AdjustTimeMinuetsToInterval(tr_FromDate.Value)
    End Sub

    Private Sub tr_ToDate_LostFocus(sender As Object, e As EventArgs) Handles tr_ToDate.LostFocus
        AdjustTimeMinuetsToInterval(tr_ToDate.Value)
    End Sub

    Public Sub LoadReportsIntoListBox(ByRef oReportGrid As DataGridView, ByVal nReportType As Integer)
        oReportGrid.Rows.Clear()
        Dim sQuery As String
        Dim oReader As OdbcDataReader
        Dim nRow As Integer

        Try
            If nReportType = ReportType.DataReport Then
                sQuery = "SELECT * FROM TBL_ReportsConfiguration WHERE AlmGroupID=" + cGroup.SelectedIndex.ToString + " AND ReportType=" + nReportType.ToString + " ORDER BY ReportTitle"
            Else
                sQuery = "SELECT * FROM TBL_ReportsConfiguration WHERE ReportType=" + nReportType.ToString + " ORDER BY ReportTitle"
            End If
            Dim oConnection As New OdbcConnection(g_sConString)
            oConnection.Open()
            Dim oCmd As New OdbcCommand(sQuery, oConnection)
            oReader = oCmd.ExecuteReader()
            While oReader.Read()
                nRow = oReportGrid.Rows.Add
                oReportGrid.Rows(nRow).Cells(0).Value = oReader("ReportID")
                oReportGrid.Rows(nRow).Cells(1).Value = oReader("ReportTitle")
                Try
                    oReportGrid.Rows(nRow).Cells(2).Value = oReader("TimeIntervalInMin")
                Catch ex As Exception
                    oReportGrid.Rows(nRow).Cells(2).Value = "1"
                End Try

            End While
            oReader.Close()
            oConnection.Close()
        Catch ex As Exception
            MsgBox(ex.Message)
            Exit Sub
        End Try
        oReportGrid.ClearSelection()
        oReportGrid.Refresh()
    End Sub

    Private Sub oTimer_Tick(sender As Object, e As EventArgs) Handles oTimer.Tick
        oTimer.Enabled = False

        If m_nReportStatusID <> 0 Then
            oProgress.Value = GetReportProgress(m_nReportStatusID)
            If oProgress.Value = 100 Then
                UpdateGrid()
            End If
        End If


        oTimer.Enabled = True
    End Sub

    Private Sub MainForm_Closing(sender As Object, e As CancelEventArgs) Handles Me.Closing
        m_bStopThread = True
    End Sub

    Private Sub UpdateGrid()
        Dim sQuery As String
        Dim oReader As OdbcDataReader
        Dim nRow As Integer

        sQuery = "SELECT * FROM tbl_reportstatus WHERE Status=3"
        oGrid.Rows.Clear()
        Try
            Using oConnection As New OdbcConnection(g_sConString)
                oConnection.Open()
                Dim oCmd As New OdbcCommand(sQuery, oConnection)
                oReader = oCmd.ExecuteReader()
                While oReader.Read()
                    nRow = oGrid.Rows.Add()
                    oGrid.Rows(nRow).Cells(0).Value = oReader("id")
                    oGrid.Rows(nRow).Cells(1).Value = oReader("outputfilename")
                    oGrid.Rows(nRow).Cells(2).Value = oReader("outputfilename")

                End While
                oConnection.Close()
            End Using
        Catch ex As Exception

        End Try

        oGrid.ClearSelection()
        oGrid.Refresh()
    End Sub

    Private Sub oGrid_CellClick(sender As Object, e As DataGridViewCellEventArgs) Handles oGrid.CellClick
        If e.RowIndex = -1 Or e.ColumnIndex <> 3 Then
            Exit Sub
        End If
        Try
            Dim sQuery As String
            sQuery = "UPDATE tbl_reportstatus SET  Status=4 WHERE id=" + oGrid.Rows(e.RowIndex).Cells(0).Value.ToString()
            ExecuteSQLInDb(sQuery)
            Process.Start(oGrid.Rows(e.RowIndex).Cells(2).Value)

        Catch ex As Exception

        End Try
        UpdateGrid()
    End Sub

    Private Sub cGroup_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cGroup.SelectedIndexChanged
        LoadReportsIntoListBox(oReportGrid, ReportType.DataReport)
    End Sub

    Private Sub bCOnfigureReports_Click(sender As Object, e As EventArgs) Handles bCOnfigureReports.Click
        Dim oDlg As New DlgReportsConfiguration
        oDlg.ShowDialog()
    End Sub

    Private Sub bConfigureAlarmReports_Click(sender As Object, e As EventArgs) Handles bConfigureAlarmReports.Click
        Dim oDlg As New DlgAlarmReportConfiguration()
        oDlg.ShowDialog()
    End Sub

    Private Sub bGenerateChart_Click(sender As Object, e As EventArgs) Handles bGenerateChart.Click

        m_nReportStatusID = InsertNewReportStatusRecord(m_nReportID, tr_FromDate.Value, tr_ToDate.Value, 1, "user Name", 1)
    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        If g_bIsBMS = 1 Then
            MactusReportLib.MactusReportLib.SynchronizeEBOPointIDNamesTable()
        ElseIf g_bIsBMS = 2 Then
            SynchronizeWirelessCDUPointIDNamesTable()
        Else
            SynchronizeIndusoftPointIDNamesTable()
        End If
        MsgBox("Data Synchronization Completed")
    End Sub

    Private Sub bConfigure_Click(sender As Object, e As EventArgs) Handles bConfigure.Click
        Dim oDlg As New DlgAppConfiguration()
        oDlg.ShowDialog()
    End Sub

    Private Sub btnExcursionReport_Click(sender As Object, e As EventArgs) Handles btnExcursionReport.Click
        Dim oDlg As New DlgExcursionConfiguration()
        oDlg.ShowDialog()
    End Sub

    Private Sub btnBatteryPercentage_Click(sender As Object, e As EventArgs) Handles btnBatteryPercentage.Click
        Dim oDlg As New DlgBatteryReportConfiguration()
        oDlg.ShowDialog()
    End Sub
End Class

File: EmsBMSReports/MainForm.cs
using System;
using System.ComponentModel;
using System.Data.Odbc;
using System.Diagnostics;
using System.Windows.Forms;
using static MactusReportLib.MactusReportLib;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System.Drawing;

namespace EmsBMSReports
{
    
    public partial class MainForm
    {

        public class GroupInfo
        {
            public int index = 0;
            public string name = "";
            public override string ToString()
            {
                return name;
            }
        }

        private ReportType m_nSelectedReportType;
        private int m_nIntervalMin = 1;
        private int m_nReportID = 0;
        private long m_nReportStatusID = 0L;

        public MainForm()
        {
            InitializeComponent();
        }

        private void MainForm_Load(object sender, EventArgs e)
        {

            try
            {
                // If g_sCurrent_login_UserID.Length > 0 Then
                string sError = "";

                var oTime = DateTime.Now;
                if (ReadDatabaseConnection(ref sError) == false)
                {
                    Interaction.MsgBox(sError);
                    Environment.Exit(0);
                }

                bCOnfigureReports.Visible = true;
                bConfigureAlarmReports.Visible = true; // g_bEnableConfiguration

                string sQuery;
                //sQuery = "UPDATE tbl_reportstatus SET status=4 where status=3";
                //ExecuteSQLInDb(sQuery);

                m_oEBOReprots = new MactusReportLib.EBOReport();
                m_oIndusoftReports = new MactusReportLib.NewInduSoftReport();

                bAreaReports.Checked = true;
                bAlarmReports.Checked = false;
                bEventReport.Checked = false;
                if (g_bIsBMS == 2)
                {
                    btnBatteryPercentage.Visible = true;
                    btnExcursionReport.Visible = true;
                }
                var argoReportGrid = oReportGrid;
                LoadReportsIntoListBox(ref argoReportGrid, (int)ReportType.DataReport);
                oReportGrid = argoReportGrid;
                cInterval.SelectedIndex = 0;
                m_nSelectedReportType = ReportType.DataReport;
                oTime = oTime.AddMilliseconds(-oTime.Millisecond);
                oTime = oTime.AddSeconds(-oTime.Second);
                oTime = oTime.AddMinutes(-oTime.Minute);
                oTime = oTime.AddHours(-oTime.Hour);
                oTime = oTime.AddDays(-7);
                tr_FromDate.Value = oTime;
                tr_ToDate.Value = oTime.AddDays(1d);
                bGenerate.Enabled = false;

                // Dim sQuery As String
                OdbcDataReader oReader;
                sQuery = "SELECT * FROM TBL_DataGroups WHERE grouptype=0 ORDER BY groupid";

                try
                {
                    using (var oConnection = new OdbcConnection(g_sConString))
                    {
                        oConnection.Open();
                        var oCmd = new OdbcCommand(sQuery, oConnection);
                        oReader = oCmd.ExecuteReader();
                        while (oReader.Read())
                        {
                            var g = new GroupInfo();
                            g.index = int.Parse(oReader["GroupId"].ToString());
                            g.name = oReader["GroupName"].ToString();
                            cGroup.Items.Add(g);
                        }
                        oConnection.Close();
                    }
                }
                catch (Exception )
                {

                }

                try
                {
                    cGroup.SelectedIndex = 0;
                    cGroup_SelectedIndexChanged(null, null);
                }
                catch (Exception )
                {

                }

                UpdateGrid();
                StartThread();


                oTimer.Enabled = true;
            }

            // Else

            // End If
            catch (Exception )
            {

            }


        }

        private void bGenerate_Click(object sender, EventArgs e)
        {
            if (m_nReportID == 0)
            {
                Interaction.MsgBox("No Report Selected");
                return;
            }
            oProgress.Value = 0;
            int nTimeInterval;

            nTimeInterval = Convert.ToInt16(cInterval.Text);

            var argdtFrom = tr_FromDate.Value;
            var argdtTo = tr_ToDate.Value;
            string argsGeneratedUserName = "user Name";
            int argnReportType = (int)m_nSelectedReportType;
            m_nReportStatusID = InsertNewReportStatusRecord(m_nReportID, ref argdtFrom, ref argdtTo, ref nTimeInterval, ref argsGeneratedUserName, ref argnReportType);
            tr_FromDate.Value = argdtFrom;
            tr_ToDate.Value = argdtTo;
            m_nSelectedReportType = (ReportType)argnReportType;
        }

        private void oReportGrid_CellClick(object sender, DataGridViewCellEventArgs e)
        {

            try
            {
                m_nReportID = Conversions.ToInteger(oReportGrid.Rows[e.RowIndex].Cells[0].Value);
                if (m_nSelectedReportType == ReportType.DataReport)
                {
                    cInterval.Text = oReportGrid.Rows[e.RowIndex].Cells[2].Value.ToString();
                    Label5.Visible = true;
                    cInterval.Visible = true;
                    m_nIntervalMin = Convert.ToInt16(oReportGrid.Rows[e.RowIndex].Cells[2].Value);
                    var argoDateTime = tr_FromDate.Value;
                    AdjustTimeMinuetsToInterval(ref argoDateTime);
                    tr_FromDate.Value = argoDateTime;
                    var argoDateTime1 = tr_ToDate.Value;
                    AdjustTimeMinuetsToInterval(ref argoDateTime1);
                    tr_ToDate.Value = argoDateTime1;
                    bGenerateChart.Enabled = true;
                }
                else
                {
                    Label5.Visible = false;
                    m_nIntervalMin = 1;
                    bGenerateChart.Enabled = false;
                }
                bGenerate.Enabled = true;
            }
            catch (Exception )
            {
                m_nReportID = 0;
                Label5.Visible = false;
                m_nIntervalMin = 1;
                bGenerate.Enabled = false;
                cInterval.Visible = false;
                bGenerateChart.Enabled = false;
            }

        }

        private void bAreaReports_CheckedChanged(object sender, EventArgs e)
        {

            cGroup.Enabled = true;
            var argoReportGrid = oReportGrid;
            LoadReportsIntoListBox(ref argoReportGrid, (int)ReportType.DataReport);
            oReportGrid = argoReportGrid;
            Label5.Visible = false;
            m_nSelectedReportType = ReportType.DataReport;
        }

        private void bAlarmReports_CheckedChanged(object sender, EventArgs e)
        {
            cGroup.Enabled = false;
            var argoReportGrid = oReportGrid;
            LoadReportsIntoListBox(ref argoReportGrid, (int)ReportType.AlarmReport);
            oReportGrid = argoReportGrid;
            Label5.Visible = false;
            m_nIntervalMin = 1;
            m_nSelectedReportType = ReportType.AlarmReport;
        }

        private void bEventReport_CheckedChanged(object sender, EventArgs e)
        {
            cGroup.Enabled = false;
            var argoReportGrid = oReportGrid;
            LoadReportsIntoListBox(ref argoReportGrid, (int)ReportType.EventReport);
            oReportGrid = argoReportGrid;
            Label5.Visible = false;
            m_nIntervalMin = 1;
            m_nSelectedReportType = ReportType.EventReport;
        }

        private void AdjustTimeMinuetsToInterval(ref DateTime oDateTime)
        {
            try
            {
                int nTemp;

                if (m_nIntervalMin == 1)
                {
                    return;
                }
                nTemp = (int)Math.Round(oDateTime.Minute / (double)m_nIntervalMin);

                nTemp = oDateTime.Minute - nTemp * m_nIntervalMin;

                oDateTime = oDateTime.AddMinutes(-nTemp);
            }
            catch (Exception )
            {

            }

        }

        private void tr_FromDate_LostFocus(object sender, EventArgs e)
        {
            var argoDateTime = tr_FromDate.Value;
            AdjustTimeMinuetsToInterval(ref argoDateTime);
            tr_FromDate.Value = argoDateTime;
        }

        private void tr_ToDate_LostFocus(object sender, EventArgs e)
        {
            var argoDateTime = tr_ToDate.Value;
            AdjustTimeMinuetsToInterval(ref argoDateTime);
            tr_ToDate.Value = argoDateTime;
        }

        public void LoadReportsIntoListBox(ref DataGridView oReportGrid, int nReportType)
        {
            oReportGrid.Rows.Clear();
            string sQuery;
            OdbcDataReader oReader;
            int nRow;

            try
            {
                if (nReportType == (int)ReportType.DataReport )
                {
                    string index = "0";
                    if (cGroup.SelectedIndex >= 0)
                        index = ((GroupInfo)(cGroup.Items[cGroup.SelectedIndex])).index.ToString();
                    sQuery = "SELECT * FROM TBL_ReportsConfiguration WHERE AlmGroupID=" + index + " AND ReportType=" + nReportType.ToString() + " ORDER BY ReportTitle";
                }
                else
                {
                    sQuery = "SELECT * FROM TBL_ReportsConfiguration WHERE ReportType=" + nReportType.ToString() + " ORDER BY ReportTitle";
                }
                var oConnection = new OdbcConnection(g_sConString);
                oConnection.Open();
                var oCmd = new OdbcCommand(sQuery, oConnection);
                oReader = oCmd.ExecuteReader();
                while (oReader.Read())
                {
                    nRow = oReportGrid.Rows.Add();
                    oReportGrid.Rows[nRow].Cells[0].Value = oReader["ReportID"];
                    oReportGrid.Rows[nRow].Cells[1].Value = oReader["ReportTitle"];
                    try
                    {
                        oReportGrid.Rows[nRow].Cells[2].Value = oReader["TimeIntervalInMin"];
                    }
                    catch (Exception )
                    {
                        oReportGrid.Rows[nRow].Cells[2].Value = "1";
                    }

                }
                oReader.Close();
                oConnection.Close();
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message);
                return;
            }
            oReportGrid.ClearSelection();
            oReportGrid.Refresh();
        }

        private void oTimer_Tick(object sender, EventArgs e)
        {
            oTimer.Enabled = false;

            if (m_nReportStatusID != 0L)
            {
                oProgress.Value = GetReportProgress(ref m_nReportStatusID);
                if (oProgress.Value == 100)
                {
                    UpdateGrid();
                }
            }


            oTimer.Enabled = true;
        }

        private void MainForm_Closing(object sender, CancelEventArgs e)
        {
            m_bStopThread = true;
        }

        private void UpdateGrid()
        {
            string sQuery = @"SELECT status.id id, 
                             status.reportId reportid,
                             rc.ReportTitle reporttitle,  
                             rc.ReportHeader reportheader, 
                             status.username username,  
                             status.fromdate fromdate,  
                             status.todate todate,  
                             status.status status,  
                             status.progress progress,  
                             status.errormessage errormessage,  
                             status.filename filename,  
                             status.outputfilename outputfilename  
                        FROM tbl_reportstatus as status  
                        Join TBL_ReportsConfiguration rc on rc.ReportID = status.reportid  
                        WHERE Status< 4";
            try
            {
                oGrid.Rows.Clear();
                using (var oConnection = new OdbcConnection(g_sConString))
                {
                    oConnection.Open();
                    var oCmd = new OdbcCommand(sQuery, oConnection);
                    var oReader = oCmd.ExecuteReader();
                    while (oReader.Read())
                    {
                        var nRow = oGrid.Rows.Add();
                        oGrid.Rows[nRow].Tag = oReader["id"];
                        oGrid.Rows[nRow].Cells[0].Value = oReader["reportheader"];
                        oGrid.Rows[nRow].Cells[1].Value = oReader["reporttitle"];
                        oGrid.Rows[nRow].Cells[2].Value = oReader["username"];
                        oGrid.Rows[nRow].Cells[3].Value = oReader["fromdate"].ToString();
                        oGrid.Rows[nRow].Cells[4].Value = oReader["todate"].ToString();
                        oGrid.Rows[nRow].Cells[5].Value = oReader["progress"].ToString()+"%";
                        if (oReader["status"].ToString() == "3")
                        {
                            DataGridViewTextBoxCell c = (DataGridViewTextBoxCell)oGrid.Rows[nRow].Cells[6];
                            c.Style.Font =  new Font(oGrid.Font,FontStyle.Underline);
                            c.Style.ForeColor = Color.Blue;
                            c.Tag = oReader["outputfilename"];
                            c.Value = "Open Report";
                            c.ToolTipText = "Open report " + c.Tag.ToString();
                            c.Value = "Open Report";
                        }
                        else
                        {
                            DataGridViewTextBoxCell c = (DataGridViewTextBoxCell)oGrid.Rows[nRow].Cells[6];
                            c.Value = "Pending";
                        }
                    }
                    oConnection.Close();
                }
            }
            catch(Exception ex)
            {
            }
            oGrid.ClearSelection();
            oGrid.Refresh();
        }

        private void oGrid_CellClick(object sender, DataGridViewCellEventArgs e)
        {
            if (e.RowIndex == -1 | e.ColumnIndex != 6)
            {
                return;
            }
            try
            {
                string sQuery;
                if (oGrid.Rows[e.RowIndex].Cells[6].Tag != null)
                {
                    sQuery = "UPDATE tbl_reportstatus SET  Status=4 WHERE id=" + oGrid.Rows[e.RowIndex].Tag.ToString();
                    ExecuteSQLInDb(sQuery);
                    Process.Start(oGrid.Rows[e.RowIndex].Cells[6].Tag.ToString());
                }
            }

            catch (Exception)
            {

            }
            UpdateGrid();
        }

        private void cGroup_SelectedIndexChanged(object sender, EventArgs e)
        {
            var argoReportGrid = oReportGrid;
            LoadReportsIntoListBox(ref argoReportGrid, (int)ReportType.DataReport);
            oReportGrid = argoReportGrid;
        }

        private void bCOnfigureReports_Click(object sender, EventArgs e)
        {
            var oDlg = new DlgReportsConfiguration();
            oDlg.ShowDialog();
        }

        private void bConfigureAlarmReports_Click(object sender, EventArgs e)
        {
            var oDlg = new DlgAlarmReportConfiguration();
            oDlg.ShowDialog();
        }

        private void bGenerateChart_Click(object sender, EventArgs e)
        {

            var argdtFrom = tr_FromDate.Value;
            var argdtTo = tr_ToDate.Value;
            int argnTimeInterval = 1;
            string argsGeneratedUserName = "user Name";
            int argnReportType = 1;
            m_nReportStatusID = InsertNewReportStatusRecord(m_nReportID, ref argdtFrom, ref argdtTo, ref argnTimeInterval, ref argsGeneratedUserName, ref argnReportType);
            tr_FromDate.Value = argdtFrom;
            tr_ToDate.Value = argdtTo;
        }

        private void Button1_Click(object sender, EventArgs e)
        {
            if (g_bIsBMS == 1)
            {
                SynchronizeEBOPointIDNamesTable();
            }
            else if (g_bIsBMS == 2)
            {
                SynchronizeWirelessCDUPointIDNamesTable();
            }
            else
            {
                SynchronizeIndusoftPointIDNamesTable();
            }
            Interaction.MsgBox("Data Synchronization Completed");
        }

        private void bConfigure_Click(object sender, EventArgs e)
        {
            var oDlg = new DlgAppConfiguration();
            oDlg.ShowDialog();
        }

        private void btnExcursionReport_Click(object sender, EventArgs e)
        {
            var oDlg = new DlgExcursionConfiguration();
            oDlg.ShowDialog();
        }

        private void btnBatteryPercentage_Click(object sender, EventArgs e)
        {
            var oDlg = new DlgBatteryReportConfiguration();
            oDlg.ShowDialog();
        }
        private void Button2_Click(object sender, EventArgs e)
        {
            var oDlg = new dlgTableView();
            oDlg.ShowDialog();
        }

        private void Button2_Click_1(object sender, EventArgs e)
        {
            var dlg = new DlgConfigureViewGroup();
            dlg.ShowDialog();
        }

        private void oGrid_CellContentClick(object sender, DataGridViewCellEventArgs e)
        {

        }

        private void cGroup_DropDown(object sender, EventArgs e)
        {
            string sQuery = "SELECT * FROM TBL_DataGroups WHERE grouptype=0 ORDER BY groupid";
            cGroup.Items.Clear();
            try
            {
                using (var oConnection = new OdbcConnection(g_sConString))
                {
                    oConnection.Open();
                    var oCmd = new OdbcCommand(sQuery, oConnection);
                    var oReader = oCmd.ExecuteReader();
                    while (oReader.Read())
                    {
                        var g = new GroupInfo();
                        g.index = int.Parse(oReader["GroupId"].ToString());
                        g.name = oReader["GroupName"].ToString();
                        cGroup.Items.Add(g);
                    }
                    oReader.Close();
                    oConnection.Close();
                }
            }
            catch (Exception)
            {

            }
        }

        private void oGrid_UserDeletedRow(object sender, DataGridViewRowEventArgs e)
        {
            string sQuery = "delete from tbl_reportstatus WHERE id=" + e.Row.Tag.ToString();
            ExecuteSQLInDb(sQuery);
        }
    }
}